components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/PreIssueAccessTokenResponseProcessor.java

```diff
@@ -154,14 +163,64 @@ private void logOperationExecutionResults(ActionType actionType,
     }
 
     @Override
-    public ActionExecutionStatus processErrorResponse(Map<String, Object> map, Event event,
-                                                      ActionInvocationErrorResponse actionInvocationErrorResponse)
+    public ActionExecutionStatus<Failure> processFailureResponse(Map<String, Object> eventContext,
+                                                        Event actionEvent,
+                                                        ActionInvocationFailureResponse failureResponse) throws
+            ActionExecutionResponseProcessorException {
+
+        handleInvalidErrorCodes(failureResponse.getFailureReason());
+        return new FailedStatus(new Failure(failureResponse.getFailureReason(),
+                failureResponse.getFailureDescription()));
+    }
+
+    /**
+     * This method validates the failedReason attribute in the FAILED status.
+     * @param errorCode
+     * @throws ActionExecutionResponseProcessorException
+     */
+    private void handleInvalidErrorCodes(String errorCode) throws ActionExecutionResponseProcessorException {
+
+        // According to the current API contract server_error is considered as an invalid value for the failureReason
+        // attribute.
+        if (isServerError(errorCode)) {
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                DiagnosticLog.DiagnosticLogBuilder diagLogBuilder = new DiagnosticLog.DiagnosticLogBuilder(
+                        ActionExecutionLogConstants.ACTION_EXECUTION_COMPONENT_ID,
+                        ActionExecutionLogConstants.ActionIDs.VALIDATE_ACTION_RESPONSE);
+                diagLogBuilder
+                        .resultMessage("Invalid value for failedReason attribute at FAILED state.")
+                        .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                        .resultStatus(DiagnosticLog.ResultStatus.FAILED)
+                        .build();
+                LoggerUtils.triggerDiagnosticLogEvent(diagLogBuilder);
+            }
+            throw new ActionExecutionResponseProcessorException("FAILED status should not be used to process" +
+                    " server errors.");
+        }
+    }
+
+    private boolean isServerError(String errorCode) {
+
+        return (errorCode.equalsIgnoreCase("internal_server_error".trim()) ||
+                errorCode.equalsIgnoreCase("server_error".trim()) ||
+                errorCode.equalsIgnoreCase("500"));
```

Comment:
Lets use HttpStatusCodes instead of 500 string

Reply:
Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2609/commits/5ff70df701d60720abd41de1bada6a7c85d4d659